pkg/controller/endpointslice/reconciler.go

```diff
@@ -0,0 +1,297 @@
+		sliceName, _ := sliceNamesToDelete.PopAny()
+		err := r.client.DiscoveryV1alpha1().EndpointSlices(service.Namespace).Delete(sliceName, &metav1.DeleteOptions{})
+		if err != nil {
+			errs = append(errs, fmt.Errorf("Error deleting %s EndpointSlice for Service %s/%s: %v", sliceName, service.Namespace, service.Name, err))
+		}
+	}
+
+	return utilerrors.NewAggregate(errs)
+}
+
+// reconcileByPortMapping compares the endpoints found in existing slices with
+// the list of desired endpoints and returns lists of slices to create, update,
+// and delete. The logic is split up into several main steps:
+// 1. Iterate through existing slices, delete endpoints that are no longer
+//    desired and update matching endpoints that have changed.
+// 2. Iterate through slices that have been modified in 1 and fill them up with
+//    any remaining desired endpoints.
+// 3. If there still desired endpoints left, try to fit them into a previously
+//    unchanged slice and/or create new ones.
+func (r *reconciler) reconcileByPortMapping(
+	service *corev1.Service,
+	existingSlices []*discovery.EndpointSlice,
+	desiredEndpoints []*discovery.Endpoint,
+	endpointMeta *endpointMeta,
+) ([]*discovery.EndpointSlice, []*discovery.EndpointSlice, sets.String) {
+	desiredSet := newEndpointSet(desiredEndpoints...)
+	slicesByName := map[string]*discovery.EndpointSlice{}
+	sliceNamesUnchanged := sets.String{}
+	sliceNamesToUpdate := sets.String{}
+	sliceNamesToDelete := sets.String{}
+
+	// 1. Iterate through existing slices to delete endpoints no longer desired
+	//    and update endpoints that have changed
+	for _, existingSlice := range existingSlices {
+		slicesByName[existingSlice.Name] = existingSlice
+		newEndpoints := []discovery.Endpoint{}
+		endpointUpdated := false
+		for _, endpoint := range existingSlice.Endpoints {
+			got := desiredSet[newEndpointHash(&endpoint)]
+			// If endpoint is desired add it to list of endpoints to keep.
+			if got != nil {
+				newEndpoints = append(newEndpoints, *got)
+				// If existing version of endpoint doesn't match desired version
+				// set endpointUpdated to ensure endpoint changes are persisted.
+				if !endpointsEqualBeyondHash(got, &endpoint) {
+					endpointUpdated = true
+				}
+				// once an endpoint has been placed/found in a slice, it is no
+				// longer desired
+				desiredSet.Delete(got)
+			}
+		}
+
+		// If an endpoint was updated or removed, mark for update or delete
+		if endpointUpdated || len(existingSlice.Endpoints) != len(newEndpoints) {
+			if len(newEndpoints) == 0 {
+				// if no endpoints desired in this slice, mark for deletion
+				sliceNamesToDelete.Insert(existingSlice.Name)
+			} else {
+				// otherwise, mark for update
+				existingSlice.Endpoints = newEndpoints
+				sliceNamesToUpdate.Insert(existingSlice.Name)
+			}
+		} else {
+			// slices with no changes will be useful if there are leftover endpoints
+			sliceNamesUnchanged.Insert(existingSlice.Name)
+		}
+	}
+
```

Comment:
Not sure if this makes it better

```
type endpointSliceTracker struct {
	endpointSliceMap map[string]*discovery.endpointSlice

	// sorted list of endpointSlice name by endpointCount
	endpointCountTracker []*discovery.endpointSlice

	needUpdate sets.String
	needDelete sets.String
	unchanged sets.String
}

func (*endpointSliceTracker t) Insert(object) {
	// insert endpointSliceMap
	// insert endpointCountTracker sequence?
}

func (*endpointSliceTracker t) Update(object) {
	// update endpointSliceMap
	// update endpointCountTracker sequence?
}

func (*endpointSliceTracker t) Delete(name)
func (*endpointSliceTracker t) Get(name)

func (*endpointSliceTracker t)  MarkForDelete(name string) error {
	if t.needUpdate.Has(name) || t.unchanged.Has(name) {
		return fmt.Errorf....
	}
	if not found  {
		return fmt.Errorf...
	}
	t.needDelete.Insert(name)
}
func (*endpointSliceTracker t)  MarkForUpdate(name string) error
func (*endpointSliceTracker t)  MarkUnchanged(name string) error

func (*endpointSliceTracker t)  ListNamesForUpdate(name string) error
func (*endpointSliceTracker t)  ListNamesForDelete(name string) error
func (*endpointSliceTracker t)  ListNamesForDelete(name string) error

type candidateFilter int 

const (
	all = iota
	updated = iota
)


func (*endpointSliceTracker t) FindInsertCandidate(numToInsert int, toleratePartialFit bool, filter candidateFilter) (object, found bool) {

	// sort the existing endpointSlice by fullness
	// filter and traverse the ednpointSlice from full to empty
	for es {
		if toleratePartialFit {
			if es still has capacity left {
				return es, true
			}

		} else {
			if es can fit numToInsert {
				return es, true
			}
		}
	}
	return nil, false
}







in the reconcile :


// try to fit in the desiredSet with endpointSlice marked for update
for desiredSet.Len() > 0 {
	candidate, ok := tracker.FindInsertCandidate(desiredSet, true, updated)
	if !ok {
		break
	}

	// fill endpoints into candidate 
	tracker.update(candidate)
}

// try to fit in the desiredSet with all endpointSlice
for desiredSet.Len() > 0 {
	candidate, ok := tracker.FindInsertCandidate(desiredSet, false, all)	
	if ok {
		// fill endpoints into candidate 
		tracker.update(candidate)	
	} else {
		create new endpoint slice
	}
}
```

Reply:
After discussing this further we decided that this would become more involved than desired here, but still worth doing in the future to help clean up all the duplicate code between proxiers.